### PR TITLE
feat(ui): allow sorting the application list (#4060)

### LINKED ISSUE #4060: Add support for sorting applications by name
# Summary
The Argo UI doesn't currently allow sorting by name or any other attribute such as last sync time.

# Motivation
I would use this when scanning the list of available apps by name, sorting by deployment time, etc.

# Proposal
Client-side sorting of the fields in memory.

### PR DESCRIPTION
**Pull request by alexef**
Screenshot 2022-12-10 at 06 57 31

https://user-images.githubusercontent.com/346576/206831880-b092f3a1-855a-4076-93db-d790c8ed4fd2.mov

replaces: #6217

fixes: #4060

Signed-off-by: Alex Eftimie <alex.eftimie@getyourguide.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by alexef** [comment]
I've tested the other use cases for Paginate (Application Details/Resource List/Grouped List) and they remain unchanged)

File: ui/src/app/applications/components/applications-list/applications-list.tsx [hunk 1]
```diff
@@ -515,6 +516,18 @@ export const ApplicationsList = (props: RouteComponentProps<{}>) => {
...
                                                                             </EmptyState>
                                                                         )}
+                                                                        sortOptions={[
+                                                                            {title: 'Name', compare: (a, b) => a.metadata.name.localeCompare(b.metadata.name)},
+                                                                            {
+                                                                                title: 'Created At',
+                                                                                compare: (a, b) => a.metadata.creationTimestamp.localeCompare(b.metadata.creationTimestamp)
+                                                                            },
+                                                                            {
+                                                                                title: 'Synchronized',
```
Comment (ashutosh16) [inline ui/src/app/applications/components/applications-list/applications-list.tsx:526]: Do you want to change `Synchronized` to `App Health`?

Comment (alexef) [reply ui/src/app/applications/components/applications-list/applications-list.tsx:526]: No, I would keep Synchronized as it is (it's the last time the sync operation finished - either successfully or not).

But an `App Health` sort order option makes sense as well: show Unknown, Degraded, Out of sync, Synced in this order!

Comment (alexef) [reply ui/src/app/applications/components/applications-list/applications-list.tsx:526]: updated

**Review by rbreeze — APPROVED** [review]
LGTM

**Comment by alexef** [comment]
@rbreeze / @alexmt can we merge this?

**Pull request merged**